src/replay_stage.rs

```diff
@@ -190,50 +190,65 @@ impl ReplayStage {
+
+            let bank = {
+                let mut bank_forks = bank_forks.write().unwrap();
+                bank_forks.set_working_bank_id(bank_id);
+                bank_forks.working_bank()
+            };
 
-        let mut current_blob_index = {
             let leader_scheduler = leader_scheduler.read().unwrap();
-            let slot = leader_scheduler.tick_height_to_slot(bank.tick_height() + 1);
+            let next_slot = leader_scheduler.tick_height_to_slot(bank.tick_height() + 1);
 
             let leader_id = leader_scheduler
-                .get_leader_for_slot(slot)
+                .get_leader_for_slot(next_slot)
                 .expect("Leader not known after processing bank");
-            trace!("node {:?} scheduled as leader for slot {}", leader_id, slot,);
+            trace!(
+                "node {:?} scheduled as leader for slot {}",
+                leader_id,
+                next_slot
+            );
+
+            // Update the working_bank() to be an unfrozen copy
+            {
+                let mut bank_forks = bank_forks.write().unwrap();
+                bank_forks.insert(bank_id, Bank::new_from_parent(&bank, &leader_id));
+            }
 
             // Send a rotation notification back to Fullnode to initialize the TPU to the right
             // state
             to_leader_sender
                 .send(TvuRotationInfo {
-                    bank: Bank::new_from_parent(&bank, &leader_id),
-                    last_entry_id: *last_entry_id.read().unwrap(),
-                    slot,
-                    leader_id,
+                    bank: bank.clone(),
+                    last_entry_id,
+                    next_slot,
+                    next_leader_id: leader_id,
                 })
                 .unwrap();
 
-            blocktree
-                .meta(slot)
-                .expect("Database error")
-                .map(|meta| meta.consumed)
-                .unwrap_or(0)
+            let bank = Arc::new(Bank::new_from_parent(&bank, &leader_id));
+
+            (
+                bank,
+                Arc::new(RwLock::new(last_entry_id)),
+                blocktree
+                    .meta(next_slot)
+                    .expect("Database error")
+                    .map(|meta| meta.consumed)
+                    .unwrap_or(0),
+            )
         };
 
         let t_replay = Builder::new()
             .name("solana-replay-stage".to_string())
             .spawn(move || {
                 let _exit = Finalizer::new(exit_.clone());
-                let mut last_leader_id = leader_scheduler_
-                    .read()
-                    .unwrap()
-                    .get_leader_for_tick(bank.tick_height() + 1)
-                    .unwrap();
                 let mut prev_slot = None;
                 let (mut current_slot, mut max_tick_height_for_slot) = {
                     let tick_height = bank.tick_height();
```

Comment:
This is potential race, as the PohService is also modifying the bank tick height by this point (signaled a rotate earlier in this function with this bank). Thus on line 256, you might get the next slot, even though you haven't processed all the entries for this one.